### PR TITLE
fix(gatsby): Multiple root query error

### PR DESCRIPTION
**Pull request by marvinjude**
## Description
This fixes  `Multiple "root" queries found` error that happens when you name a query in page template

## Related Issues
[sc-54658]

Fixes #36420

### CODE REVIEW

File: packages/gatsby/src/query/__tests__/__snapshots__/query-compiler.js.snap [hunk 1]
```diff
@@ -153,37 +158,7 @@ Object {
...
-  11 | }",
-      "name": "AnotherQuery",
-      "otherName": "mockFileQuery",
-    },
-    "filePath": "mockFile",
-    "id": "85910",
-  },
-]
-`;
+exports[`actual compiling errors on double root 1`] = `Array []`;
```
Comment (pieh) [inline packages/gatsby/src/query/__tests__/__snapshots__/query-compiler.js.snap:161]: this indicate that we might have gone too liberal with changes so far - updating snapshots with new fields is one thing, not throwing expected errors is another :)

Comment (marvinjude) [reply packages/gatsby/src/query/__tests__/__snapshots__/query-compiler.js.snap:161]: Yeah we  lose error in cases like this probably others too 

```
export const query = graphql`
  query DemoPage {
    site {
      id
    }
  }
  query DemoPage {
    site {
      id
    }
  }
`;
```

**Review by tyhopp — COMMENTED** [review]
Does this remove the limitation mentioned in https://github.com/gatsbyjs/gatsby/discussions/36420#discussioncomment-3428987? If so we should update the docs in this PR as well

File: packages/gatsby-plugin-mdx/README.md [hunk 2]
```diff
@@ -664,41 +664,8 @@ However, we'd recommend placing such files into the `src/pages` directory and `g
 
 1. Instead of querying for the `body` on the MDX node, the page template now receives the transformed MDX as `children` property.
 1. You no longer need to use `<MDXRenderer>` as you can use `{children}` directly.
-1. If you have given your query a name (e.g. `query PostTemplate`) you have to remove the name. Gatsby automatically creates a name internally, but if a name is provided you'll see a "Duplicate query" warning.
 
-```diff
```
Comment (LekoArts) [inline packages/gatsby-plugin-mdx/README.md:669]: You removed too much here. It still has to show the `children` diff noted in the first bullet point.

You also shouldn't change this to four backticks, you've made the whole section below a code block

File: packages/gatsby/src/query/query-compiler.js [hunk 3]
```diff
@@ -303,31 +310,9 @@ const processDefinitions = ({
...
-      )
-
-      store.dispatch(
-        actions.queryExtractionGraphQLError({
-          componentPath: filePath,
-        })
-      )
-      continue
-    }
+    const { filePath, templatePath } = operation
```
Comment (LekoArts) [inline packages/gatsby/src/query/query-compiler.js:315]: Move this destructuring to the top of this scope, also destructure `def` and then you can replace `operation.def` with `def`

File: packages/gatsby/src/query/query-compiler.js [hunk 4]
```diff
@@ -396,15 +382,44 @@ const processDefinitions = ({
       continue
     }
 
+    const printedDocument = print(document)
+    // Check for duplicate page/static queries in the same component.
+    // (config query is not a duplicate of page/static query in the component)
+    // TODO: make sure there is at most one query type per component (e.g. one config + one page)
```
Comment (LekoArts) [inline packages/gatsby/src/query/query-compiler.js:388]: This TODO still a TODO?

File: packages/gatsby/src/query/query-compiler.js [hunk 5]
```diff
@@ -396,15 +382,44 @@ const processDefinitions = ({
...
       name,
-      text: print(document),
+      text: printedDocument,
       originalText: originalDefinition.text,
       path: filePath,
       isHook: originalDefinition.isHook,
       isStaticQuery: originalDefinition.isStaticQuery,
       isConfigQuery: originalDefinition.isConfigQuery,
       // ensure hash should be a string and not a number
+      templatePath,
```
Comment (LekoArts) [inline packages/gatsby/src/query/query-compiler.js:422]: The comment one line above is at the wrong place

**Review by LekoArts — CHANGES_REQUESTED** [review]

File: packages/gatsby/src/query/query-compiler.js (hunk 4 above)
Comment (pieh) [reply packages/gatsby/src/query/query-compiler.js:388]: Yup, we didn't touch this part in this PR

Comment (pieh) [reply packages/gatsby/src/query/query-compiler.js:388]: This comment also is not exactly relevant to what this PR is doing, so it should stay

**Review by LekoArts — APPROVED** [review]

**Pull request merged**